### PR TITLE
[#2488] Change STUDENTS and TEAMS to *_EXCLUDING_SELF for feedback paths

### LINKED ISSUE #2488: Instructor: edit questions: give more recipient type options
Suggested by a user:
My suggestion is about the “feedback recipient” field. As it stands, we have this useful option “Giver’s team members and Giver”, which allows for simultaneous self- and peer-assessment. However, there is no equivalent option for the other types of recipient. In other words, I think you should add the following options:

·         Other students in the course, and Giver

·         Other teams in the course, and Giver’s team

### PR DESCRIPTION
**Pull request by xyliew25**
Part of #2488

**Outline of Solution**

As suggested, 
1. For feedback path, modify the code such that the previously unused `STUDENTS_EXCLUDING_SELF` enum indicates exactly what it is named, but keep `STUDENTS` similarly defined.
2. Write and run a data migration script to migrate all `STUDENTS` to `STUDENTS_EXCLUDING_SELF`.
3. Remove the code that interprets `STUDENTS` as "students excluding self".

### CODE REVIEW

**Comment by xyliew25** [comment]
As of now, I have attempted Step 1 and 2.

Also, I understand that Issue [#2448](https://github.com/TEAMMATES/teammates/issues/2488) targets `TEAMS` as well, but I would like to get the `STUDENTS` part reviewed and finalized first before attempting the `TEAMS` part, as I believe similar implementation applies.

**Comment by wkurniawan07** [comment]
> As of now, I have attempted Step 1 and 2.

You cannot do step 3 now anyway; not until this set of change has been merged and the script run in the production DB.

**Review by wkurniawan07 — COMMENTED** [review]
Implementation is sufficient. Remember to update the script and the backend logic to also involve `TEAMS_INCLUDING_SELF`.

File: src/web/services/feedback-questions.service.ts [hunk 1]
```diff
@@ -61,22 +61,22 @@ export class FeedbackQuestionsService {
       case FeedbackQuestionType.RANK_RECIPIENTS:
       case FeedbackQuestionType.CONSTSUM_RECIPIENTS:
         paths.set(FeedbackParticipantType.SELF,
-          [FeedbackParticipantType.STUDENTS, FeedbackParticipantType.STUDENTS_IN_SAME_SECTION,
-            FeedbackParticipantType.INSTRUCTORS, FeedbackParticipantType.TEAMS,
+          [FeedbackParticipantType.STUDENTS_EXCLUDING_SELF, FeedbackParticipantType.STUDENTS_IN_SAME_SECTION,
+            FeedbackParticipantType.INSTRUCTORS, FeedbackParticipantType.TEAMS_EXCLUDING_SELF,
```
[the remark this answers is not in the document]
Comment (xyliew25) [reply src/web/services/feedback-questions.service.ts:65]: Ah yes, will do.

**Comment by xyliew25** [comment]
> Implementation is sufficient. Remember to update the script and the backend logic to also involve `TEAMS_INCLUDING_SELF`.

Oh wait, not sure if I got the wrong idea, but did you mean `TEAMS_EXCLUDING_SELF`?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/client/java/teammates/client/scripts/DataMigrationForStudentsAndTeamsRecipientType.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        DataMigrationEntitiesBaseScript<FeedbackQuestion> {
+
+    public static void main(String[] args) {
+        new DataMigrationForStudentsAndTeamsRecipientType().doOperationRemotely();
+    }
+
+    @Override
+    protected Query<FeedbackQuestion> getFilterQuery() {
+        return ofy().load().type(FeedbackQuestion.class)
+                .filter("recipientType =", null);
```
[the remark this answers is not in the document]
Comment (xyliew25) [reply src/client/java/teammates/client/scripts/DataMigrationForStudentsAndTeamsRecipientType.java:24]: I overlooked it, my bad.

File: src/web/app/pages-instructor/instructor-session-edit-page/__snapshots__/instructor-session-edit-page.component.spec.ts.snap [hunk 3]
```diff
@@ -4039,7 +4039,7 @@ exports[`InstructorSessionEditPageComponent should snap with new question added
                   <div
                     class="offset-1 col-md-4 text-md-right"
                   >
-                     The maximum number of students each respondent should give feedback to: 
+                     The maximum number of students_excluding_self each respondent should give feedback to:
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-session-edit-page/__snapshots__/instructor-session-edit-page.component.spec.ts.snap:4042]: This unfortunately is no longer user-friendly. If previously we can just lowercase (I assume), we may either need a dedicated pipe here, or maybe just trimming the `_EXCLUDING_SELF` at the end.

File: src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java [hunk 4]
```diff
@@ -50,7 +50,7 @@ public enum FeedbackParticipantType {
     /**
      * Teams of the course, excluding the response giver.
      */
-    TEAMS_EXCLUDING_SELF(false, false, false),
+    TEAMS_EXCLUDING_SELF(false, true, true),
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackParticipantType.java:53]: I don't think the third boolean should be `true`

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/web/app/components/teammates-common/feedback-path.pipe.ts [hunk 5]
```diff
@@ -68,3 +72,33 @@ export class RecipientTypeDescriptionPipe implements PipeTransform {
...
+ */
+@Pipe({
+  name: 'recipientTypeSimplifiedDescription',
+})
+export class RecipientTypeSimplifiedDescriptionPipe implements PipeTransform {
+
+  /**
+   * Transforms {@link FeedbackParticipantType} to a simple description from recipient's perspective.
+   */
+  transform(type: FeedbackParticipantType): any {
```
[the remark this answers is not in the document]
Comment (xyliew25) [reply src/web/app/components/teammates-common/feedback-path.pipe.ts:87]: Ah yes, what about the existing `GiverTypeDescriptionPipe` and `RecipientTypeDescriptionPipe`? Their return types are also `any`. Should I change them as well, or leave them?

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/web/app/components/teammates-common/feedback-path.pipe.ts:87]: Leave them, they are irrelevant for this issue/PR

**Review by samuelfangjw — APPROVED** [review]
LGTM! Great work on this :)

**Pull request merged**